hibernate-core/src/main/java/org/hibernate/cache/spi/support/NaturalIdReadWriteAccess.java

```diff
@@ -74,7 +74,7 @@ public boolean afterInsert(SharedSessionContractImplementor session, Object key,
 		try {
 			writeLock().lock();
 			Lockable item = (Lockable) getStorageAccess().getFromCache( key, session );
-			if ( item == null ) {
+			if ( item == null || item.getValue() != value) {
```

Comment:
Rather than checking the item's value I think the cache entry should not have been created in the first place. Accessing the cache after the `persist`/`remove` in a single transaction would lead to incorrectly finding an entry.

It would be great if you could check why the first transaction creates it even though the entity is deleted and see if we can avoid it.

Reply:
>>It would be great if you could check why the first transaction creates it even though the entity is deleted 

Thanks for reviewing this. You can find an explanation of what's happening in the [Jira Issue](https://hibernate.atlassian.net/browse/HHH-16557) (second comment).
I don't remember exactly why on inserts the storage in NaturaIdCache were deferred to afterTransactionCompletion.
I only can immagine that in the moment of insert (persist-call) the naturalId fields might not yet be initialized/set to their final values and therefore it was decided to deffer it to afterTransactionCompletion.
Im pretty sure that when we not deffer the insert to afterTransactionCompletion some other test will fail.
I know the proposed solution is not very neat, but I don't know how to do it better.

>>Accessing the cache after the persist/remove in a single transaction would lead to incorrectly finding an entry.
Not exactly. Only internally we will produce a cache hit, resolving the naturalId values to the entities primary key.
Then when hiberante tries to resolve the entity itself by the PK it recognizes that it does not exist,
so from user point of view nothing is broken.

Reply:
The cache item is saved after the transaction completes to make sure we create it only if the insert was successful. You can see this in `NaturalIdResolutionsImpl`, where `NaturalIdDataAccess#afterInsert` is called.

That would be a great starting point to start trying to see if we can avoid the cache item creation in the first place.

Reply:
Well there might be different approaches, each own with its pros and cons

1. Before processing the InsertActionQueue, compare it's elements with the DeleteActionQueue and annihilate creations and deletions for the same entity. 

   Pros: avoids useles cache item creation/eviction not only in NaturalIdCache but also in entity shared cache.  
   Cons: might have performance impact and other side-effects.

2. Deffer also delete of NaturalIdCache eviction to afterDelete (afterTransactionCompletion).

Reply:
@mbladel Overthinkig these 3 alternatives, I think that alternative n.2 is the best, most straight forward approach.